tests/test_project.py

```diff
@@ -1128,7 +1128,7 @@ def mock_project(self, project_class=None):
 
 class TestExecutionDynamicProject(TestExecutionProject):
     project_class = _DynamicTestProject
-    expected_number_of_steps = 10
+    expected_number_of_steps = 7
```

Comment:
I think this was a bug in the tests (see comment above for explanation of the incorrect postcondition). This should take seven steps. In each step, the scheduler makes a pass and then the project runs `submit(...)`.

1. Scheduler: no jobs. Project: submit op1, op2, op3 for all jobs (those three operations are eligible at the beginning).
2. Scheduler: move op1-3 jobs from submitted to holding. Project: no action.
3. Scheduler: move op1-3 jobs from holding to queued. Project: no action.
4. Scheduler: execute op1-3 jobs, set status to inactive. Project: submit op4 for all jobs (eligible because op1 is complete).
5. Scheduler: move op4 jobs from submitted to holding. Project: no action.
6. Scheduler: move op4 jobs from holding to queued. Project: no action.
7. Scheduler: execute op4 jobs, set status to inactive. Project: no action (all done).

Reply:
Agreed, this makes sense. And the old behavior just happened to work because `test_resubmit` runs exactly three more steps than necessary (to flush things out of the queue), which happens to be the value originally included. Without an upper bound that test would run forever.